README.md
# pangocairocffi

[![Latest PyPi Release](https://img.shields.io/pypi/v/pangocairocffi.svg)](https://pypi.python.org/pypi/pangocairocffi)
[![Supported Python Versions](https://img.shields.io/pypi/pyversions/pangocairocffi.svg?style=flat)](https://pypi.python.org/pypi/pangocairocffi)
[![Build Status](https://github.com/leifgehrmann/pangocairocffi/actions/workflows/build.yml/badge.svg)](https://github.com/leifgehrmann/pangocairocffi/actions)
[![Documentation Status](https://readthedocs.org/projects/pangocairocffi/badge/?version=latest)](https://pangocairocffi.readthedocs.io/en/latest/?badge=latest)
[![Code Coverage](https://codecov.io/gh/leifgehrmann/pangocairocffi/branch/master/graph/badge.svg)](https://codecov.io/gh/leifgehrmann/pangocairocffi)

pangocairocffi is a [CFFI]-based set of Python bindings for the
[cairo rendering methods with pango]. It is meant to be used in
conjunction with [cairocffi] and [pangocffi].

[CFFI]: https://cffi.readthedocs.org/
[pangocffi]: https://github.com/leifgehrmann/pangocffi
[cairocffi]: https://cairocffi.readthedocs.io/en/stable/
[cairo rendering methods with pango]: https://docs.gtk.org/PangoCairo/

The bindings are currently not fully implemented. Feel free to make a pull
request to contribute!

## Installation and usage

See ['Overview'] for information on how to install the necessary libraries.

See ['Python API Reference'] for additional information on all the objects.

['Overview']: https://pangocairocffi.readthedocs.io/en/latest/overview.html
['Python API Reference']: https://pangocairocffi.readthedocs.io/en/latest/modules.html

## Contributing

If you would like to contribute to this project, either by leaving feedback or
submitting a pull request, please read '[CONTRIBUTING.md]'.

[CONTRIBUTING.md]: CONTRIBUTING.md


NEWS.rst
Changelog
---------

Version 0.6.0
.............

Released on 2020-12-30.

* Added two new static methods to ``PangoCairoFontMap``:

    * ``set_default``
    * ``get_default``

Version 0.5.0
.............

Released on 2020-12-19.

* Added the ability to configure loading a specific library via the environment
  variable ``PANGOCAIRO_LOCATION``.

Version 0.4.0
.............

Released on 2020-11-13.

**Breaking Changes**

* C-FFI bindings are now generated at runtime, rather than at installation.
  This was done to avoid common installation issues like
  ``ModuleNotFoundError: No module named 'pangocairocffi._generated'``. This
  change does mean the bindings are re-compiled the first time
  ``import pangocairocffi`` is called in a python session, but it takes less
  than a second to do this. If there are any issues with this change, please
  raise an issue in the issue tracker.
* pangocairocffi now depends on pangocffi v0.8.0. Older versions are not
  compatible.
* Support for Python 3.5 has been dropped because it has reached end-of-life.

Version 0.3.2
.............

Released on 2020-10-06.

* Reverted previous unsuccessful patch.

Version 0.3.1
.............

Released on 2020-10-06.

* Hopefully fixed issue with ``pangocairocffi`` not being cache-able as a wheel
  by pip.

Version 0.3.0
.............

Released on 2019-10-08.

* Extended library names to include current Win64 library names.

Version 0.2.2 - 0.2.6
.....................

Released on 2019-03-23.

These were a series of changes that happened rapidly since the issues were not
easy to replicate manually.

- Version 0.2.2

  - Upgraded dependency of ``pangocffi`` from 0.3.0 to 0.4.0.
  - Possible fix for SandboxViolation when installing via easy_install/
    setuptools

- Version 0.2.3

  - Possible fix for installation issues related to ``ffi_pango.py`` not being
    installed in the correct directory, and also the ``pangocffi`` dependency
    not being installed in the setup process.

- Version 0.2.4

  - ``ffi_pango.py`` still was not being installed correctly. So now rather
    than relying on the file being generated, the file exists hardcoded in the
    repository.

- Version 0.2.5

  - setuptools apparently does not follow the ``package_data`` rule.
    A ``MANIFEST.in`` file has been added to fix this.
    (see https://stackoverflow.com/a/14159430)

- Version 0.2.6

  - ``include_package_data`` and ``zip_safe`` need to be set.

Version 0.2.2
.............

Released on 2019-03-23.

Version 0.2.1
.............

Released on 2019-03-21.

- Improved linting coverage by refactoring the cffi build script.

Version 0.2.0
.............

Released on 2019-03-19.

- Upgraded dependency of ``pangocffi`` from 0.1.1 to 0.3.0.
  This gave us the ability to implement ``show_glyph_item()``

- Replaced incorrect type hinting of ``ctypes`` with ``ffi.CData``

- Added new tests, including examples in the documentation:

  - ``test_error_underline.py``

  - ``test_extents.py``

  - ``test_glyph_item.py``

Version 0.1.0
.............

Released on 2019-03-09.

First PyPI release.


tests/test_font_map.py
from pangocairocffi import PangoCairoFontMap


def test_font_map():
    font_map_default = PangoCairoFontMap.get_default()
    font_map_new = PangoCairoFontMap()
    assert font_map_new != font_map_default

    resolution = font_map_new.get_resolution()
    assert resolution == 96
    font_map_new.set_resolution(123)
    assert font_map_new.get_resolution() == 123

    PangoCairoFontMap.set_default(font_map_new)
    assert font_map_new == PangoCairoFontMap.get_default()
    PangoCairoFontMap.set_default(None)


pangocairocffi/__init__.py
"""
    pangocairocffi
    ~~~~~~~~~

    CFFI-based pangocairo bindings for Python
    See README for details.

"""
import os
import warnings
from typing import List
import ctypes.util
from .ffi_build import ffi
import cairocffi


def _dlopen(dl_name: str, generated_ffi, names: List[str]):
    """
    :param dl_name:
        The name of the dynamic library. This is also used to determine the
        environment variable name to lookup. For example, if dl_name is
        "pangocairo", this function will attempt to load "PANGOCAIRO_LOCATION".
    :param generated_ffi:
        The FFI for pangocairo, generated by pangocairocffi.
    :param names:
        An array of library names commonly used across different platforms.
    :return:
        A FFILibrary instance for the library.
    """

    # Try environment locations if set
    env_location = os.getenv(f'{dl_name.upper()}_LOCATION')
    if env_location:
        names.append(env_location)
        try:
            return generated_ffi.dlopen(env_location)
        except OSError:
            warnings.warn(f"dlopen() failed to load {dl_name} library:"
                          f" '{env_location}'. Falling back.")

    # Try various names for the same library, for different platforms.
    for name in names:
        for lib_name in (name, 'lib' + name):
            try:
                path = ctypes.util.find_library(lib_name)
                lib = generated_ffi.dlopen(path or lib_name)
                if lib:
                    return lib
            except OSError:
                pass
    raise OSError(
        f"dlopen() failed to load {dl_name} library: {' / '.join(names)}"
    )


pangocairo = _dlopen('pangocairo', ffi, ['pangocairo-1.0', 'pangocairo-1.0-0'])


# Imports are normally always put at the top of the file.
# But the wrapper API requires that the pango library be loaded first.
# Therefore, we have to disable linting rules for these lines.
# from .constants import *  # noqa

# from .pango_cairo_font_map import PangoCairoFontMap

def _get_cairo_t_from_cairo_ctx(cairo_context: cairocffi.Context):
    # Cairo does not give access to the pointer publicly. So this is the best
    # we can do.
    # noinspection PyProtectedMember
    return cairo_context._pointer  # noqa


from .render_functions import *  # noqa
from .create_update_functions import *  # noqa
from .font_functions import *  # noqa
from .font_map import PangoCairoFontMap  # noqa


pangocairocffi/ffi_build.py
"""
    pangocairocffi.ffi_build
    ~~~~~~~~~~~~~~~~~~~~~~~~

    Build the cffi bindings
"""

import sys
from pathlib import Path
from cffi import FFI
from pangocffi import ffi as ffi_pango

sys.path.append(str(Path(__file__).parent))

# Create an empty _generated folder if needed
(Path(__file__).parent / '_generated').mkdir(exist_ok=True)

# Read the CFFI definitions
c_definitions_cairo_file = open(
    str(Path(__file__).parent / 'c_definitions_cairo.txt'),
    'r'
)
c_definitions_cairo = c_definitions_cairo_file.read()
c_definitions_pangocairo_file = open(
    str(Path(__file__).parent / 'c_definitions_pangocairo.txt'),
    'r'
)
c_definitions_pangocairo = c_definitions_pangocairo_file.read()

# cffi definitions, in the order outlined in:
ffi = FFI()
ffi.include(ffi_pango)
ffi.set_source('pangocairocffi._generated.ffi', None)
ffi.cdef(c_definitions_cairo)
ffi.cdef(c_definitions_pangocairo)

if __name__ == '__main__':
    ffi.compile()


pangocairocffi/font_functions.py
import ctypes
from typing import Optional
from pangocffi import Context
from . import pangocairo
from . import ffi


# Todo: pangocairo.Font needs to be implemented
# def get_cairo_scaled_font_pointer(font: Font) -> ffi.CData:
#     """
#     Gets the cairo_scaled_font_t used by font . The scaled font can be
#     referenced and kept using ``cairo_scaled_font_reference()``.
#
#     :param font:
#         a Pango font from a Pango-Cairo font map
#     :return:
#         the cairo_scaled_font_t used by ``font``
#     """
#     return pangocairo.pango_cairo_font_get_scaled_font(font.get_pointer())


def set_resolution(context: Context, dpi: float) -> None:
    """
    Sets the resolution for the context. This is a scale factor between points
    specified in a PangoFontDescription and Cairo units. The default value is
    96, meaning that a 10 point font will be 13 units high.
    (10 * 96. / 72. = 13.3).

    :param context:
        a Pango context
    :param dpi:
        the resolution in "dots per inch". (Physical inches aren't actually
        involved; the terminology is conventional.) A 0 or negative value
        means to use the resolution from the font map.
    """
    pangocairo.pango_cairo_context_set_resolution(context.get_pointer(), dpi)


def get_resolution(context: Context) -> float:
    """
    Returns the resolution for the Pango context.

    :param context:
        a Pango context
    :return:
        the resolution in "dots per inch". A negative value will be returned
        if no resolution has previously been set.
    """
    return pangocairo.pango_cairo_context_get_resolution(context.get_pointer())


def set_font_options(
        context: Context,
        options: Optional[ffi.CData]
) -> None:
    """
    Sets the font options used when rendering text with this context.
    These options override any options that pango_cairo_update_context()
    derives from the target surface.

    :param context:
        a Pango context
    :param options:
        a cairo_font_options_t, or ``None`` to unset any previously set
        options.
    """
    if options is None:
        options = ffi.NULL
    context_pointer = context.get_pointer()
    pangocairo.pango_cairo_context_set_font_options(context_pointer, options)


def get_font_options(context: Context) -> Optional[ffi.CData]:
    """
        Retrieves any font rendering options previously set with
        pango_cairo_context_set_font_options(). This function does not report
        options that are derived from the target surface by
        pango_cairo_update_context()

        :param context:
            a Pango Context
        :return:
            a cairo_font_options_t pointer previously set on the context,
            otherwise ``None``.
    """
    context_pointer = context.get_pointer()
    font_option_pointer = pangocairo.pango_cairo_context_get_font_options(
        context_pointer
    )
    if font_option_pointer == ffi.NULL:
        return None
    return font_option_pointer


pangocairocffi/render_functions.py
import cairocffi
import pangocffi
from . import pangocairo
from . import ffi
from . import _get_cairo_t_from_cairo_ctx


# Todo: Font and GlyphString needs to be implemented
# def show_glyph_string(
#         cairo_context: cairocffi.Context,
#         font: pangocffi.Font,
#         glyphs: pangocffi.GlyphString
# ) -> None:
#     """
#     Draws the glyphs in glyphs in the specified cairo context. The origin of
#     the glyphs (the left edge of the baseline) will be drawn at the current
#     point of the cairo context.
#
#     :param cairo_context:
#         a Cairo context
#     :param font:
#         a Pango font from a PangoCairoFontMap
#     :param glyphs:
#         a Pango glyph string
#     """
#     cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
#     pangocairo.pango_cairo_show_glyph_string(
#         cairo_context_pointer,
#         font.get_pointer(),
#         glyphs.get_pointer()
#     )


def show_glyph_item(
        cairo_context: cairocffi.Context,
        text: str,
        glyph_item: pangocffi.GlyphItem
) -> None:
    """
    Draws the glyphs in ``glyph_item`` in the specified cairo context,
    embedding the text associated with the glyphs in the output if the output
    format supports it (PDF for example), otherwise it acts similar to
    :meth:`show_glyph_string()`.

    The origin of the glyphs (the left edge of the baseline) will be drawn at
    the current point of the cairo context.

    Note that ``text`` is the start of the text for layout, which is then
    indexed by ``glyph_item->item->offset``.

    :param cairo_context:
        a Cairo context
    :param text:
        the UTF-8 text that ``glyph_item`` refers to
    :param glyph_item:
        a Pango glyph item
    """
    cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    text_pointer = ffi.new('char[]', text.encode('utf8'))
    pangocairo.pango_cairo_show_glyph_item(
        cairo_context_pointer,
        text_pointer,
        glyph_item.get_pointer()
    )


# Todo: LayoutLine needs to be implemented
# def show_layout_line(
#         cairo_context: cairocffi.Context,
#         line: pangocffi.LayoutLine
# ) -> None:
#     """
#     Draws a PangoLayoutLine in the specified cairo context. The origin of the
#     glyphs (the left edge of the line) will be drawn at the current point of
#     the cairo context.
#
#     :param cairo_context:
#         a Cairo context
#     :param line:
#         a Pango layout line
#     """
#     cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
#     pangocairo.pango_cairo_show_layout_line(
#         cairo_context_pointer,
#         line.get_pointer()
#     )


def show_layout(
        cairo_context: cairocffi.Context,
        layout: pangocffi.Layout
) -> None:
    """
    Draws a Pango Layout in the specified cairo context. The top-left corner
    of the PangoLayout will be drawn at the current point of the cairo context.

    :param cairo_context:
        a Cairo context
    :param layout:
        a Pango layout
    """
    cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    pangocairo.pango_cairo_show_layout(
        cairo_context_pointer,
        layout.get_pointer()
    )


def show_error_underline(
        cairo_context: cairocffi.Context,
        x: float,
        y: float,
        width: float,
        height: float
) -> None:
    """
    Draw a squiggly line in the specified cairo context that approximately
    covers the given rectangle in the style of an underline used to indicate
    a spelling error. (The width of the underline is rounded to an integer
    number of up/down segments and the resulting rectangle is centered in the
    original rectangle)

    :param cairo_context:
        a Cairo context
    :param x:
        The X coordinate of one corner of the rectangle
    :param y:
        The Y coordinate of one corner of the rectangle
    :param width:
        Non-negative width of the rectangle
    :param height:
        Non-negative height of the rectangle
    """
    cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    pangocairo.pango_cairo_show_error_underline(
        cairo_context_pointer,
        x,
        y,
        width,
        height
    )


# Todo: Font and GlyphString needs to be implemented
# def glyph_string_path(
#         cairo_context: cairocffi.Context,
#         font: pangocffi.Font,
#         glyphs: pangocffi.GlyphString
# ) -> None:
#     """
#     Adds the glyphs in glyphs to the current path in the specified cairo
#     context. The origin of the glyphs (the left edge of the baseline) will
#     be at the current point of the cairo context.
#
#     :param cairo_context:
#         a Cairo context
#     :param font:
#         a Pango font from a PangoCairoFontMap
#     :param glyphs:
#         a Pango glyph string
#     """
#     cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
#     pangocairo.pango_cairo_glyph_string_path(
#         cairo_context_pointer,
#         font.get_pointer(),
#         glyphs.get_pointer()
#     )


# Todo: LayoutLine needs to be implemented
# def layout_line_path(
#     """
#     Adds the text in PangoLayoutLine to the current path in the specified
#     cairo context. The origin of the glyphs (the left edge of the line) will
#     be at the current point of the cairo context.
#
#     :param cairo_context:
#         a Cairo context
#     :param line:
#         a Pango layout line
#     """
#     cairo_context: cairocffi.Context,
#     line: pangocffi.LayoutLine
# ) -> None:
#     cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
#     pangocairo.pango_cairo_layout_line_path(
#         cairo_context_pointer,
#         line.get_pointer()
#     )


def layout_path(
        cairo_context: cairocffi.Context,
        layout: pangocffi.Layout
) -> None:
    """
    Adds the text in a ``Pango.Layout`` to the current path in the specified
    cairo context. The top-left corner of the ``Pango.Layout`` will be at the
    current point of the cairo context.

    :param cairo_context:
        a Cairo context
    :param layout:
        a Pango layout
    """
    cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    pangocairo.pango_cairo_layout_path(
        cairo_context_pointer,
        layout.get_pointer()
    )


def error_underline_path(
        cairo_context: cairocffi.Context,
        x: float,
        y: float,
        width: float,
        height: float
) -> None:
    """
    Add a squiggly line to the current path in the specified cairo context that
    approximately covers the given rectangle in the style of an underline used
    to indicate a spelling error. (The width of the underline is rounded to an
    integer number of up/down segments and the resulting rectangle is centered
    in the original rectangle)

    :param cairo_context:
        a Cairo context
    :param x:
        The X coordinate of one corner of the rectangle
    :param y:
        The Y coordinate of one corner of the rectangle
    :param width:
        Non-negative width of the rectangle
    :param height:
        Non-negative height of the rectangle
    """
    cairo_context_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    pangocairo.pango_cairo_error_underline_path(
        cairo_context_pointer,
        x,
        y,
        width,
        height
    )


pangocairocffi/create_update_functions.py
import cairocffi
import pangocffi
from . import pangocairo
from . import _get_cairo_t_from_cairo_ctx


def create_context(cairo_context: cairocffi.Context) -> pangocffi.Context:
    """
    Creates a context object set up to match the current transformation and
    target surface of the Cairo context. This context can then be used to
    create a layout using ``pangocffi.Layout()``.

    This function is a convenience function that creates a context using the
    default font map, then updates it to ``cairo_context``. If you just need to
    create a layout for use with ``cairo_context`` and do not need to access
    PangoContext directly, you can use :meth:`create_layout()` instead.

    :param cairo_context:
        a Cairo context
    :return:
        a Pango context
    """
    cairo_t_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    layout_pointer = pangocairo.pango_cairo_create_context(cairo_t_pointer)
    return pangocffi.Context.from_pointer(layout_pointer, gc=True)


def update_context(
        cairo_context: cairocffi.Context,
        pango_context: pangocffi.Context
) -> None:
    """
    Updates a PangoContext previously created for use with Cairo to match the
    current transformation and target surface of a Cairo context. If any
    layouts have been created for the context, it's necessary to call
    pango_layout_context_changed() on those layouts.

    :param cairo_context:
        a Cairo context
    :param pango_context:
        a Pango context, from a pango-cairo font map
    """
    cairo_t_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    pangocairo.pango_cairo_update_context(
        cairo_t_pointer,
        pango_context.get_pointer()
    )


def create_layout(cairo_context: cairocffi.Context) -> pangocffi.Layout:
    """
    Creates a layout object set up to match the current transformation and
    target surface of the Cairo context. This layout can then be used for text
    measurement with functions like ``get_size()`` or drawing with functions
    like :meth:`show_layout()`. If you change the transformation or target
    surface for ``cairo_context``, you need to call
    :meth:`update_layout()`

    This function is the most convenient way to use Cairo with Pango, however
    it is slightly inefficient since it creates a separate PangoContext object
    for each layout. This might matter in an application that was laying out
    large amounts of text.

    :param cairo_context:
        a Cairo context
    :return:
        a Pango layout
    """
    cairo_t_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    layout_pointer = pangocairo.pango_cairo_create_layout(cairo_t_pointer)
    return pangocffi.Layout.from_pointer(layout_pointer, gc=True)


def update_layout(
        cairo_context: cairocffi.Context,
        layout: pangocffi.Layout
) -> None:
    """
    Updates the private Pango ``Context`` of a Pango ``Layout`` created with
    :meth:`create_layout()` to match the current transformation and target
    surface of a Cairo context.

    :param cairo_context:
        a Cairo context
    :param layout:
        a Pango layout
    """
    cairo_t_pointer = _get_cairo_t_from_cairo_ctx(cairo_context)
    pangocairo.pango_cairo_update_layout(cairo_t_pointer, layout.get_pointer())


pangocairocffi/font_map.py
from typing import Optional

from . import pangocairo, ffi
from pangocffi import Context
from pangocffi import pango


class PangoCairoFontMap:
    """
    API not *fully* implemented yet.

    **Todo:** This class should extend FontMap from pangocffi once it is
    implemented. This class in theory should be able to inherit the functions
    listed here:
    https://developer.gnome.org/pango/stable/pango-Fonts.html, with the prefix
    ``pango_font_map_X``. For example: ``create_context``, ``load_font``,
    ``load_fontset``, ``list_families``.

    ``PangoCairoFontMap`` is an interface exported by font maps for use with
    Cairo. The actual type of the font map will depend on the particular font
    technology Cairo was compiled to use.
    """

    def __init__(self):
        """
        ``__init__`` Creates a new PangoCairoFontMap object; a ``fontmap`` is
        used to cache information about available fonts, and holds certain
        global parameters such as the resolution. In most cases, you can use
        :py:meth:`PangoCairoFontMap.get_default()` instead.

        Note that the type of the returned object will depend on the particular
        font backend Cairo was compiled to use; You generally should only use
        the ``PangoFontMap`` and ``PangoCairoFontMap`` interfaces on the
        returned object.

        You can override the type of backend returned by using an environment
        variable ``PANGOCAIRO_BACKEND``. Supported types, based on your build,
        are fc (fontconfig), win32, and coretext. If requested type is not
        available, NULL is returned. I.E. this is only useful for testing,
        when at least two backends are compiled in.
        """
        self._init_pointer(pangocairo.pango_cairo_font_map_new())

    def _init_pointer(self, pointer: ffi.CData):
        pointer = ffi.cast('PangoCairoFontMap *', pointer)
        self._pointer = pointer

    def get_pointer(self) -> ffi.CData:
        """
        Returns the pointer to the font map

        :return:
            the pointer to the font map.
        """
        return self._pointer

    @classmethod
    def from_pointer(cls, pointer: ffi.CData) -> 'PangoCairoFontMap':
        """
        Instantiates a :class:`PangoCairoFontMap` from a pointer.

        :return:
            the pango-cairo font map.
        """
        if pointer == ffi.NULL:
            raise ValueError('Null pointer')
        self = object.__new__(cls)
        cls._init_pointer(self, pointer)
        return self

    @classmethod
    def get_default(
            cls,
    ) -> 'PangoCairoFontMap':
        """
        Gets a default ``PangoCairoFontMap`` to use with Cairo.

        Note that the type of the returned object will depend on the particular
        font backend Cairo was compiled to use; You generally should only use
        the PangoFontMap and PangoCairoFontMap interfaces on the returned
        object.

        The default Cairo fontmap can be changed by using
        :py:meth:`PangoCairoFontMap.set_default()`. This can be used to change
        the Cairo font backend that the default fontmap uses for example.

        Note that since Pango 1.32.6, the default fontmap is per-thread. Each
        thread gets its own default fontmap. In this way, PangoCairo can be
        used safely from multiple threads.

        :return:
            the default PangoCairo fontmap for the current thread. This object
            is owned by Pango and must not be freed.
        """
        font_map_pointer = pangocairo.pango_cairo_font_map_get_default()
        return cls.from_pointer(font_map_pointer)

    @classmethod
    def set_default(
            cls,
            fontmap: Optional['PangoCairoFontMap'] = None
    ) -> None:
        """
        Sets a default PangoCairoFontMap to use with Cairo.

        This can be used to change the Cairo font backend that the default
        fontmap uses for example. The old default font map is unreffed and the
        new font map referenced.

        Note that since Pango 1.32.6, the default fontmap is per-thread. This
        function only changes the default fontmap for the current thread.
        Default fontmaps of existing threads are not changed. Default fontmaps
        of any new threads will still be created using
        pango_cairo_font_map_new().

        A value of ``None`` for fontmap will cause the current default font
        map to be released and a new default font map to be created on demand,
        using ``pango_cairo_font_map_new()``.

        :return:
            the pango-cairo font map.
        """
        fontmap_pointer = ffi.NULL
        if fontmap is not None:
            fontmap_pointer = fontmap.get_pointer()
        pangocairo.pango_cairo_font_map_set_default(fontmap_pointer)

    @classmethod
    def from_cairo_font_type(
            cls,
            cairo_font_type_pointer: ffi.CData
    ) -> 'PangoCairoFontMap':
        """
        Instantiates a :class:`PangoCairoFontMap` from a Cairo context.

        :return:
            the pango-cairo font map.
        """
        font_map_pointer = pangocairo.pango_cairo_font_map_new_for_font_type(
            cairo_font_type_pointer
        )
        return cls.from_pointer(font_map_pointer)

    def __eq__(self, other):
        if isinstance(other, PangoCairoFontMap):
            return self.get_pointer() == other.get_pointer()
        return NotImplemented

    def get_cairo_font_type_pointer(self) -> ffi.CData:
        """
        Returns the pointer to the type of Cairo font backend that ``fontmap``
        uses

        :return:
            the pointer to the ``cairo_font_type_t``.
        """
        return pangocairo.pango_cairo_font_map_get_font_type(
            self.get_pointer()
        )

    def set_resolution(self, dpi: float) -> None:
        """
        Sets the resolution for the ``fontmap``. This is a scale factor between
        points specified in a Pango FontDescription and Cairo units. The
        default value is 96, meaning that a 10 point font will be 13
        units high. (10 * 96. / 72. = 13.3).

        :param dpi:
             the resolution in "dots per inch". (Physical inches aren't
             actually involved; the terminology is conventional.)
        """
        pangocairo.pango_cairo_font_map_set_resolution(self.get_pointer(), dpi)

    def get_resolution(self) -> float:
        """
        Returns the resolution for the ``fontmap``.

        :return:
            the resolution in "dots per inch"
        """
        return pangocairo.pango_cairo_font_map_get_resolution(
            self.get_pointer()
        )

    def create_context(self) -> Context:
        """
        Creates a Pango ``Context`` connected to ``fontmap``. This is
        equivalent to ``pango.Context()`` followed by
        ``context.set_font_map()``.

        :return:
            the newly allocated Pango ``Context``.
        """
        context_pointer = pango.pango_font_map_create_context(
            self.get_pointer()
        )
        return Context.from_pointer(context_pointer, gc=True)
